python-package/xgboost/spark/core.py

```diff
@@ -1092,12 +1093,87 @@ def _get_feature_col(
             )
         return features_col, feature_col_names
 
+    def _get_pred_contrib_col_name(self) -> str:
+        """Return the pred_contrib_col col name"""
+        pred_contrib_col_name = ""
+        if (
+            self.isDefined(self.pred_contrib_col)
+            and self.getOrDefault(self.pred_contrib_col) != ""
+        ):
+            pred_contrib_col_name = self.getOrDefault(self.pred_contrib_col)
+
+        return pred_contrib_col_name
+
+    def _out_schema(self) -> str:
+        """The returned type of the user-defined function. The value can be either a
+        :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string."""
+
+        return (
+            f"{pred.prediction} double, {pred.pred_contrib} array<double>"
+            if self._get_pred_contrib_col_name()
+            else "double"
+        )
+
+    def _get_predict_func(self) -> Callable:
+        """Return the true prediction function which will be running on the executor side"""
+
+        predict_params = self._gen_predict_params_dict()
+        pred_contrib_col_name = self._get_pred_contrib_col_name()
+
+        def _predict(
+            model: XGBModel, X: ArrayLike, base_margin: Optional[np.ndarray]
+        ) -> Union[pd.DataFrame, pd.Series]:
+            data = {}
+            preds = model.predict(
+                X,
+                base_margin=base_margin,
+                validate_features=False,
+                **predict_params,
+            )
+            data[pred.prediction] = pd.Series(preds)
+
+            result = data[pred.prediction]
+            if pred_contrib_col_name:
+                contribs = pred_contribs(model, X, base_margin)
+                data[pred.pred_contrib] = pd.Series(list(contribs))
+                result = pd.DataFrame(data=data)
+
+            return result
+
+        return _predict
+
+    def _post_transform(self, dataset: DataFrame, pred_col: Column) -> DataFrame:
+        """Post process of transform"""
+        prediction_col_name = self.getOrDefault(self.predictionCol)
+        single_pred = "," not in self._out_schema()
```

Comment:
That's a bit, hmm, unconventional. Is there a way to refactor this to make it more "standard"?

Reply:
The _out_schema has typing hint with str returned which is a DDL formatted string, so if there're many columns, it must have at least a ",". So it's conventional when we check if it is a single column according to if  "," is in the schema.

Reply:
We are meta-programming by manipulating strings here.

Reply:
Maybe just use the `pred_contrib_col_name` as a predicate?

Reply:
fixed this issue.